pkg/services/ngalert/notifier/channels/pushover.go

```diff
@@ -0,0 +1,246 @@
+		return nil, fmt.Errorf("failed to convert OK priority to integer: %w", err)
+	}
+	retry, _ := strconv.Atoi(model.Settings.Get("retry").MustString())
+	expire, _ := strconv.Atoi(model.Settings.Get("expire").MustString())
+	alertingSound := model.Settings.Get("sound").MustString()
+	okSound := model.Settings.Get("okSound").MustString()
+	uploadImage := model.Settings.Get("uploadImage").MustBool(true)
+
+	if userKey == "" {
+		return nil, alerting.ValidationError{Reason: "user key not found"}
+	}
+	if APIToken == "" {
+		return nil, alerting.ValidationError{Reason: "API token not found"}
+	}
+	return &PushoverNotifier{
+		NotifierBase: old_notifiers.NewNotifierBase(&models.AlertNotification{
+			Uid:                   model.UID,
+			Name:                  model.Name,
+			Type:                  model.Type,
+			DisableResolveMessage: model.DisableResolveMessage,
+			Settings:              model.Settings,
+			SecureSettings:        model.SecureSettings,
+		}),
+		UserKey:          userKey,
+		APIToken:         APIToken,
+		AlertingPriority: alertingPriority,
+		OKPriority:       okPriority,
+		Retry:            retry,
+		Expire:           expire,
+		Device:           device,
+		AlertingSound:    alertingSound,
+		OKSound:          okSound,
+		Upload:           uploadImage,
+		Message:          model.Settings.Get("message").MustString(`{{ template "default.message" .}}`),
+		Title:            model.Settings.Get("title").MustString(`{{ template "default.title" . }}`),
+		tmpl:             t,
+		log:              log.New("alerting.notifier.pushover"),
+	}, nil
+}
+
+// Notify sends an alert notification to Slack.
+func (pn *PushoverNotifier) Notify(ctx context.Context, as ...*types.Alert) (bool, error) {
+	headers, uploadBody, err := pn.genPushoverBody(ctx, as...)
+	if err != nil {
+		pn.log.Error("Failed to generate body for pushover", "error", err)
+		return false, err
+	}
+
+	cmd := &models.SendWebhookSync{
+		Url:        PUSHOVERENDPOINT,
+		HttpMethod: "POST",
+		HttpHeader: headers,
+		Body:       uploadBody.String(),
+	}
+
+	if err := bus.DispatchCtx(ctx, cmd); err != nil {
+		pn.log.Error("Failed to send pushover notification", "error", err, "webhook", pn.Name)
+		return false, err
+	}
+
+	return true, nil
+}
+func (pn *PushoverNotifier) SendResolved() bool {
+	return !pn.GetDisableResolveMessage()
+}
+
+func (pn *PushoverNotifier) genPushoverBody(ctx context.Context, as ...*types.Alert) (map[string]string, bytes.Buffer, error) {
+	var b bytes.Buffer
+
+	u, err := url.Parse(pn.tmpl.ExternalURL.String())
+	if err != nil {
+		return nil, b, fmt.Errorf("failed to parse ")
+	}
+	u.Path = path.Join(u.Path, "/alerting/list")
+	ruleURL := u.String()
+
+	alerts := types.Alerts(as...)
+
+	var tmplErr error
+	data := notify.GetTemplateData(ctx, pn.tmpl, as, gokit_log.NewNopLogger())
+	tmpl := notify.TmplText(pn.tmpl, data, &tmplErr)
+
+	w := multipart.NewWriter(&b)
+	// set a specific boundary to be used by the tests
+	err = w.SetBoundary(BOUNDARY)
```

Comment:
I am not sure if this is safe for non tests. Usually boundary is long and random to avoid clashes. I would suggest renaming it to `BoundaryForTest` with default as `""`, and only set it here if `BoundaryForTest != ""`. (In the test you can set something but also reset it to `""` as a cleanup).